components/datetime/src/provider/helpers.rs

```diff
@@ -121,15 +127,50 @@ impl DateTimePatterns for provider::gregory::DatePatternsV1 {
         Ok(Pattern::from_bytes_combination(s, date, time)?)
     }
 
-    fn get_pattern_for_time_length(&self, length: length::Time) -> Result<Pattern> {
-        let time = &self.time;
-        let s = match length {
+    /// Look up the proper pre-computed pattern for a given length. If a preference for an hour
+    /// cycle is set, it will look look up a pattern in the time_h11_12 or time_h23_h24 provider
+    /// data, and then manually modify the symbol in the pattern if needed.
+    fn get_pattern_for_time_length(
+        &self,
+        length: length::Time,
+        preferences: &Option<preferences::Bag>,
+    ) -> Result<Pattern> {
+        // Apply the hour cycle preference if it exists.
+        if let Some(preferences::Bag {
+            hour_cycle: Some(hour_cycle_pref),
+        }) = preferences
+        {
+            let time = match hour_cycle_pref {
+                preferences::HourCycle::H11 | preferences::HourCycle::H12 => &self.time_h11_h12,
+                preferences::HourCycle::H23 | preferences::HourCycle::H24 => &self.time_h23_h24,
+            };
+            let mut pattern = Pattern::from_bytes(match length {
+                length::Time::Full => &time.full,
+                length::Time::Long => &time.long,
+                length::Time::Medium => &time.medium,
+                length::Time::Short => &time.short,
+            })?;
+            for item in pattern.items_mut() {
+                if let PatternItem::Field(fields::Field { symbol, length: _ }) = item {
+                    if let fields::FieldSymbol::Hour(_) = symbol {
+                        *symbol = fields::FieldSymbol::Hour(hour_cycle_pref.field());
+                    }
+                }
+            }
+            return Ok(pattern);
+        }
+
+        let time = match self.preferred_hour_cycle {
+            crate::pattern::CoarseHourCycle::H11H12 => &self.time_h11_h12,
+            crate::pattern::CoarseHourCycle::H23H24 => &self.time_h23_h24,
+        };
+
+        Ok(Pattern::from_bytes(match length {
             length::Time::Full => &time.full,
             length::Time::Long => &time.long,
             length::Time::Medium => &time.medium,
             length::Time::Short => &time.short,
-        };
-        Ok(Pattern::from_bytes(s)?)
+        })?)
     }
 }
```

Comment:
`suggestion (n-b)`: The function would read better for me if it was chunked into three general steps:
1) Get `time` from either prefs of preferred_hour_cycle
2) Parse it into a pattern
3) If prefs are set, override the hour field.

I haven't try to write it, and the function is small enough, that if you prefer to keep it as-is, I'm not worried about maintenance, so non-blocking suggestion.